### PR TITLE
add rays and rings restful api

### LINKED ISSUE #39: hugegraph在进行出度遍历时存在性能问题
描述：从一个已知节点出发，向外（out）遍历，直至不存在叶子节点为止
语句：g.V().hasId(\"xxx\").repeat(out(\"look\")).until(outE(\"look\").count().is(0)).path()
数据：3000多条边+3000多个顶点；  **上述语句执行花了12s**+  请问什么原因 可否优化

### PR DESCRIPTION
**Pull request by zhoney**
fix #39

Change-Id: I41f82b73cee0ce657b3bcaaf1a68cf3479788fbf

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rays.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        HugeTraverser traverser = new HugeTraverser(g);
+        MultivaluedMap<Boolean, HugeTraverser.Path> paths =
+                       traverser.subGraphPaths(source, dir, edgeLabel, depth,
+                                               degree, capacity, limit);
+        List<HugeTraverser.Path> pathList = paths.get(false);
+        if (pathList == null) {
+            pathList = ImmutableList.of();
+        }
+        return manager.serializer(g).writePaths("ray_paths", pathList, false);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rays.java]: "ray_paths" => "rays" or "paths"

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rays.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                  graph, sourceV, direction, edgeLabel, depth, degree, limit);
+
+        Id source = VertexAPI.checkAndParseVertexId(sourceV);
+        Directions dir = Directions.convert(EdgeAPI.parseDirection(direction));
+
+        HugeGraph g = graph(manager, graph);
+
+        HugeTraverser traverser = new HugeTraverser(g);
+        MultivaluedMap<Boolean, HugeTraverser.Path> paths =
+                       traverser.subGraphPaths(source, dir, edgeLabel, depth,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rays.java]: don't collect rings which will be ignored

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rings.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+        HugeTraverser traverser = new HugeTraverser(g);
+        MultivaluedMap<Boolean, HugeTraverser.Path> paths =
+                       traverser.subGraphPaths(source, dir, edgeLabel, depth,
+                                               degree, capacity, limit);
+        List<HugeTraverser.Path> pathList = paths.get(true);
+        if (pathList == null) {
+            pathList = ImmutableList.of();
+        }
+        return manager.serializer(g).writePaths("ring_paths", pathList,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rings.java]: "ring_paths" => "rings" or "paths"

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rings.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                  graph, sourceV, direction, edgeLabel, depth, degree, limit);
+
+        Id source = VertexAPI.checkAndParseVertexId(sourceV);
+        Directions dir = Directions.convert(EdgeAPI.parseDirection(direction));
+
+        HugeGraph g = graph(manager, graph);
+
+        HugeTraverser traverser = new HugeTraverser(g);
+        MultivaluedMap<Boolean, HugeTraverser.Path> paths =
+                       traverser.subGraphPaths(source, dir, edgeLabel, depth,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/traversers/Rings.java]: ditto

File: hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/Serializer.java [hunk 5]
```diff
@@ -65,5 +67,9 @@
     public String writePaths(String name, Collection<HugeTraverser.Path> paths,
                              boolean withCrossPoint);
 
+    public String writeSubGraphPaths(
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/serializer/Serializer.java]: remove unused interface

File: hugegraph-api/src/main/java/com/baidu/hugegraph/version/ApiVersion.java [hunk 6]
```diff
@@ -72,10 +72,11 @@
      *
      * version 0.8:
      * [0.28] Issue-153: Add task-cancel API
+     * [0.29] Issue-39: Add sub-graph paths RESTful API
      */
 
     // The second parameter of Version.of() is for IDE running without JAR
-    public static final Version VERSION = Version.of(ApiVersion.class, "0.28");
+    public static final Version VERSION = Version.of(ApiVersion.class, "0.29");
```
Comment (Linary) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/version/ApiVersion.java]: remember modify Implementation-Version in pom

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java [hunk 7]
```diff
@@ -555,6 +595,103 @@ public boolean reachLimit() {
...
+            MultivaluedMap<Id, Node> newVertices = newMultivalueMap();
+            Iterator<Edge> edges;
+            // Traversal vertices of previous level
+            for (List<Node> nodes : this.sources.values()) {
+                for (Node n : nodes) {
+                    edges = edgesOfVertex(n.id(), direction,
+                                          this.label, this.degree);
+                    if (!edges.hasNext()) {
+                        // Rays found
+                        if (rings) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java]: this.rings

File: hugegraph-api/src/main/java/com/baidu/hugegraph/version/ApiVersion.java [hunk 8]
```diff
@@ -72,10 +72,11 @@
      *
      * version 0.8:
      * [0.28] Issue-153: Add task-cancel API
+     * [0.29] Issue-39: Add sub-graph paths RESTful API
```
Comment (Linary) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/version/ApiVersion.java]: Add rays and rings RESTful API

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java [hunk 9]
```diff
@@ -555,6 +595,103 @@ public boolean reachLimit() {
...
+            List<Path> paths = new ArrayList<>();
+            MultivaluedMap<Id, Node> newVertices = newMultivalueMap();
+            Iterator<Edge> edges;
+            // Traversal vertices of previous level
+            for (List<Node> nodes : this.sources.values()) {
+                for (Node n : nodes) {
+                    edges = edgesOfVertex(n.id(), direction,
+                                          this.label, this.degree);
+                    if (!edges.hasNext()) {
+                        // Rays found
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java]: Reach the end, rays found

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java [hunk 10]
```diff
@@ -138,6 +138,46 @@ public HugeTraverser(HugeGraph graph) {
...
+                                     int maxDepth, long degree, long capacity,
+                                     long limit, boolean rings) {
+        E.checkNotNull(sourceV, "source vertex id");
+        E.checkNotNull(dir, "direction");
+        checkPositive(maxDepth, "Paths max depth");
+        checkDegree(degree);
+        checkCapacity(capacity);
+        checkLimit(limit);
+
+        List<Path> paths = new ArrayList<>();
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java]: move to line 170

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java [hunk 11]
```diff
@@ -555,6 +595,103 @@ public boolean reachLimit() {
...
+                        continue;
+                    }
+                    while (edges.hasNext()) {
+                        HugeEdge edge = (HugeEdge) edges.next();
+                        Id target = edge.id().otherVertexId();
+                        this.accessedVertices.add(target);
+
+                        if (n.contains(target)) {
+                            // Rings found
+                            if (!rings) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java]: this.rings

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java [hunk 12]
```diff
@@ -555,6 +595,103 @@ public boolean reachLimit() {
...
+                            paths.add(new Path(null, prePath));
+                            this.pathCount++;
+                            if (reachLimit()) {
+                                return paths;
+                            }
+                            continue;
+                        }
+
+                        // Add node to next start-nodes
+                        newVertices.add(target, new Node(target, n));
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeTraverser.java]: should be added only if it's not a cycle path?

**Pull request merged**